File: raspberry/SmartHome/thirdpartydevices/src/main/java/raspberry/smarthome/thirdpartydevices/xiaomi/gateway/device/WaterLeakSensor.java
package raspberry.smarthome.thirdpartydevices.xiaomi.gateway.device;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Optional;

import static smarthome.library.common.constants.DeviceTypes.WATER_LEAK_SENSOR_TYPE;

public class WaterLeakSensor extends ExtendedDevice {

    private OnWaterLeakListener listener;

    public WaterLeakSensor(String sid) {
        super(sid, WATER_LEAK_SENSOR_TYPE);
    }

    public WaterLeakSensor(String sid, OnWaterLeakListener listener) {
        this(sid);
        setListener(listener);
    }

    public void setListener(OnWaterLeakListener listener) {
        this.listener = listener;
    }

    @Override
    public void parseData(String cmd) {
        try {
            JSONObject o = new JSONObject(cmd);

            if(!o.isNull(STATUS_KEY)) {
                setStatus(o.getString(STATUS_KEY));

                if(getStatus().equals(STATUS_WATER_LEAK))
                    Optional.ofNullable(listener).ifPresent(OnWaterLeakListener::onWaterLeak);

                else if(getStatus().equals(STATUS_NO_WATER_LEAK))
                    Optional.ofNullable(listener).ifPresent(OnWaterLeakListener::onNoWaterLeak);
            }

            if(!o.isNull(VOLTAGE_KEY)) {
                String vol = o.getString(VOLTAGE_KEY);
                setVoltage(Float.parseFloat(vol) / 1000);
            }

        } catch (JSONException | NumberFormatException e) {
            e.printStackTrace();
        }
    }

    public interface OnWaterLeakListener {

        void onWaterLeak();

        void onNoWaterLeak();
    }
}

File: raspberry/SmartHome/thirdpartydevices/src/main/java/smarthome/raspberry/thirdpartydevices/xiaomi/new_gateway/device/WaterLeakSensor.kt
package smarthome.raspberry.thirdpartydevices.xiaomi.new_gateway.device

import org.json.JSONException
import org.json.JSONObject
import smarthome.library.common.constants.DeviceTypes.WATER_LEAK_SENSOR_TYPE
import smarthome.library.common.constants.GATEWAY_WATER_LEAK_CONTROLLER
import smarthome.raspberry.thirdpartydevices.xiaomi.new_gateway.constants.STATUS_KEY
import smarthome.raspberry.thirdpartydevices.xiaomi.new_gateway.constants.STATUS_WATER_LEAK
import smarthome.raspberry.thirdpartydevices.xiaomi.new_gateway.controller.VoltageController
import smarthome.raspberry.thirdpartydevices.xiaomi.new_gateway.controller.WaterLeakController
import smarthome.raspberry.thirdpartydevices.xiaomi.new_gateway.controller.listeners.WaterLeakListener

class WaterLeakSensor(sid: String,
                      waterLeakListener: WaterLeakListener? = null)
    : Device(sid, WATER_LEAK_SENSOR_TYPE, waterLeakListener = waterLeakListener) {

    private var leak: Boolean = false

    init {
        addControllers(WaterLeakController(this),
                VoltageController(this))
    }

    override fun parseData(json: String) {
        try {
            val o = JSONObject(json)

            if (!o.isNull(STATUS_KEY)) {
                getControllerByType(GATEWAY_WATER_LEAK_CONTROLLER).state = o.getString(STATUS_KEY)

                val leakStatus = getControllerByType(GATEWAY_WATER_LEAK_CONTROLLER).state == STATUS_WATER_LEAK

                if(leak != leakStatus) {
                    waterLeakListener?.onWaterLeakStatusChanged(leakStatus)
                    leak = leakStatus
                }
            }

            setVoltage(o)

        } catch (e: JSONException) {
            reportDataParseError(e)
        }
    }
}